content/blog/open-sourcing-workshops-on-demand-part3-understanding-the-backend.md

```diff
@@ -0,0 +1,340 @@
+
+* T﻿he first API call will update the password data for the participant's allocated student.
+* T﻿he second API call will update the the participant's allocated student's status to active.
+
+T﻿hese changes will trigger on the frontend web portal application the sending of the second email to the participant.  This email will contain the necessary information for the participant to connect to its notebooks environment. The participant will then run the workshop. For each workshop, a dedicated time window is allocated. Some workshops will take longer to be run than others. The time windows varies from 2 to 4 hours maximum. The workshops are somehow time bombed. This means that a the very moment, the participant hit the register button on the frontend web portal, the cloak starts ticking. Some background checks take place on the web portal to verify time spent since the registration to a given workshop. As a consequence, a reminder email is sent a hour before the finish line. When the bell rings at the end of the class, a new procmail API call is made to the backend server ordering a **CLEANUP** action. The particpant can also trigger this action by registering to a new workshop before the end of the current one. He will have to provide the necessary information to the frontend web portal in order to end the current workshop. 
+
+L﻿et's see what is happening on the backend server to perform this **CLEANUP** scenario.
+
+![](/img/wod-blogserie3-cleanup.png "backend server CLEANUP workflow")
+
+A﻿s you can see, it does not differ much from the **CREATE**. We still need to gather data to interact with the proper workshop from the right student. The .procmail.rc is providing us these infos. Then, the automation kicks in through `procmail-action-sh` script.
+
+T﻿he verb is now **CLEANUP**. As a consequence, step4 is now **CLEANUP**.
+
+N﻿othing changes from 5 to 9.
+
+1﻿0- `get_wod_completion_ratio()` allows us to retrieve through a simple computing of the numbers of notebooks cells executed thoughout the different exercices of the workshop a ratio. This enables us to see how much of the workshop is actually run. Participants are asked to fill out a form in a conclusion notebook which is present in every student's workshop's folder. But only 10 % of participant do fill it. That leaves many participants for which we don't know whether they liked or not or more importantly, did they actually run it ? 
+
+T﻿his completion ratio script provides us this data and we store it in our database. 
+
+1﻿1- API call to send the completion ratio figure to the database. This can be later queried to build up nice reporting dashboard as explained here by my colleague Didier Lalli in the following [article](https://developer.hpe.com/blog/open-source-elasticsearch-helped-us-globally-support-virtual-labs/).
```

Comment:
Add the word "a" prior to "nice reporting". Place a comma after "colleague" and another comma after "Lalli".

Reply:
done